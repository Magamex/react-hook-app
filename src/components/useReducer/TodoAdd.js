import React from 'react'
import { useForm } from '../../Hooks/useForm';

export const TodoAdd = ({ handleAdd }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
          return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };
      
        handleAdd(newTodo);
        reset();
    }

  return (
    <div>
        <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="tarea"
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
        />
        <button type="submit">Agregar</button>
      </form>
    </div>
  )
}
