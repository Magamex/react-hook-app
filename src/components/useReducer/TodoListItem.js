import React from "react";

export const TodoListItem = ({ todo, i, handleToggle, handleDelete }) => {
  return (
    <li>
      <p
        className={todo.done && "tachado"}
        onClick={() => handleToggle(todo.id)}
      >
        {i + 1}. {todo.desc}
      </p>
      <button onClick={() => handleDelete(todo.id)}>Borrar</button>
    </li>
  );
};
