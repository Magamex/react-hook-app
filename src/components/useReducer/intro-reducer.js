const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = (state = initialState, action) => {
    
    if(action?.type === 'agregar'){
        return [...state, action.payload];
    }
    
    return state;
}

const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo
}

let todos = todoReducer(initialState, agregarTodoAction)

console.log(todos);