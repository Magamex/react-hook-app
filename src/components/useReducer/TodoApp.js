import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";

import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

import "../../styles.css";
const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
  // return [{
  //     id: new Date().getTime(),
  //     desc: 'Hola mundo',
  //     done: false
  // }]
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = {
      type: "delete",
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: "toggle",
      payload: todoId,
    });
  };

  const handleAdd = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  }

  return (
    <div>
      <h1>TodoApp ({todos.length})</h1>
      <TodoAdd handleAdd={handleAdd}/>
      <hr />

      <TodoList
        todos={todos}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    </div>
  );
};
