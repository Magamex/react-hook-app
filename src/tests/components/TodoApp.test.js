import React from 'react';
import { TodoApp } from './../../components/useReducer/TodoApp';
import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react'

const demoTodos = [{
    id:1,
    desc: 'Aprender React',
    done: false
},{
    id:2,
    desc: 'Aprender React Native',
    done: false
}];

describe('Pruebas en <TodoApp/>', () => {

    const wrapper = shallow(<TodoApp/>);
    Storage.prototype.setItem = jest.fn(()=>{});

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de agregar un TODO', () => {
        const wrapper = mount(<TodoApp/>);

        act(()=>{
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[1]);
        })

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('Debe de eliminar un TODO', () => {
        wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (1)');

        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)');
    });

});