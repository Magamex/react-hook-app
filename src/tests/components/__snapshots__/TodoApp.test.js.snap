// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <TodoAdd
    handleAdd={[Function]}
  />
  <hr />
  <TodoList
    handleDelete={[Function]}
    handleToggle={[Function]}
    todos={Array []}
  />
</div>
`;
